<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">

<head>
<title>README.html</title>

</head>

<body>

<h1 id="preseq">Preseq</h1>
<p><em>preseq</em> is an empirical Bayes method for predicting the molecular complexity of sequencing libraries or samples on the basis of data from very shallow sequencing runs.</p>
<em>In english</em>: provide preseq with either a <strong>BED</strong> or <strong>BAM</strong> format and it will give you the following information:
<ul>
<li>
yield estimates from experiment and subsampled experiments
</li>
<li>
predicted yield estimates from additional sequencing
</li>
<li>
predict the genomic coverage given current reads
</li>
</ul>
<h2 id="basic-usage">Basic Usage</h2>
<em>c_curve</em>: generates a complexity plot of the genomic library in BED or BAM format
<pre><code>$ ./preseq c_curve -o complexity_output.txt input.bed</code></pre>
<em>lc_extrap</em>: estimate the future yield of a genomic library using an initial experiment
<pre><code>$ ./preseq lc_extrap -o future_yield.txt input.bed</code></pre>
<em>gc_extrap</em>: predicts the genomic coverage from deep sequencing based on the initial sample
<pre><code>$ ./preseq gc_extrap -o future_coverage.txt input.mr</code></pre>
<p>
<strong>Note</strong>: for additional information please refer to the <a href="http://smithlabresearch.org/wp-content/uploads/manual.pdf">manual</a>
</p>

</body>
</html>
